tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,536 @@
+      return scope.release(block.get());
+    }
+  }
+
+  /**
+   * Runs {@code block} and releases and returns the result, then closes any <b>other</b> tensors created during its
+   * execution (or attached to the scope).
+   * <p>Tensors can be released using the passed scope.
+   *
+   * @return the released result of {@code block}
+   */
+  public static <T extends HasTensors> T produceHasTensorsWithCleanup(Function<TensorScope, T> block) {
+    try (TensorScope scope = new TensorScope()) {
+      return scope.release(block.apply(scope));
+    }
+  }
+
+
+  /**
+   * Runs {@code block} and releases and returns the result, then closes any <b>other</b> tensors created during its
+   * execution.
+   * <p>To release other tensors, use {@link #produceTensorWithCleanup(Function)}.
+   *
+   * @return the released result of {@code block}
+   */
+  public static <T extends Iterable<? extends Tensor>> T produceTensorsWithCleanup(Supplier<T> block) {
+    try (TensorScope scope = new TensorScope()) {
+      return scope.release(block.get());
+    }
+  }
+
+  /**
+   * Runs {@code block} and releases and returns the result, then closes any <b>other</b> tensors created during its
+   * execution (or attached to the scope).
+   * <p>Tensors can be released using the passed scope.
+   *
+   * @return the released result of {@code block}
+   */
+  public static <T extends Iterable<? extends Tensor>> T produceTensorsWithCleanup(Function<TensorScope, T> block) {
+    try (TensorScope scope = new TensorScope()) {
+      return scope.release(block.apply(scope));
+    }
+  }
+
+  /**
+   * Create a new tensor scope.  If {@code autoAttach} is false, will not automatically manage tensors.
+   *
+   * @see TensorScope
+   */
+  TensorScope() {
+    this.parent = currentScope();
+    currentScope.set(this);
+
+    if (this.parent != null) {
+      synchronized (this.parent) {
+        this.parent.children.add(this);
+      }
+    }
+  }
+
+  /**
+   * Closes this scope and its tensors, and any inner scopes.
+   */
+  @Override
+  public synchronized void close() {
+    if (closed) {
+      return;
+    }
+
+    children.forEach(TensorScope::close);
+    tensors.forEach(Tensor::close);
+
+    closed = true;
+
+    if (parent != null) {
+      parent.children.remove(this);
+    }
+
+    if (currentScope() == this) {
+      currentScope.set(this.parent);
+    }
+  }
+
+  /**
+   * Release the tensors and child scopes of this scope to it's parent, <b>without closing them</b>.
+   * <p>
+   * Semantically, calling this method makes all of the resources in this scope the parent's responsibility, as if this
+   * scope had never existed.
+   * <p>
+   * This will close this scope, but does not close any of it's resources.
+   *
+   * @throws IllegalStateException if this scope has no parent. If this happens, * the scope is not closed and no
+   * resources are released.
+   */
+  public synchronized void releaseToParent() {
```

Comment:
Since there is a `release` method to release a single tensor, renaming this method and the following to `releaseAllToParent()`  and `releaseAll()` respectively would clarify a bit more their intent.